File: epi_judge_python/matrix_rotation.py
from typing import List

from test_framework import generic_test


def rotate(matrix, k, n):
    to = [(k, i) for i in range(k, k + n - 1)]
    ri = [(i, k + n - 1) for i in range(k, k + n - 1)]
    bo = [(k + n - 1, i) for i in range(k + n - 1, k, -1)]
    le = [(i, k) for i in range(k + n - 1, k, -1)]
    for a, b, c, d in zip(to, ri, bo, le):
        (
            matrix[b[0]][b[1]],
            matrix[c[0]][c[1]],
            matrix[d[0]][d[1]],
            matrix[a[0]][a[1]],
        ) = (
            matrix[a[0]][a[1]],
            matrix[b[0]][b[1]],
            matrix[c[0]][c[1]],
            matrix[d[0]][d[1]],
        )


def rotate_matrix(square_matrix: List[List[int]]) -> None:
    n = len(square_matrix)
    for i in range(len(square_matrix) // 2):
        rotate(square_matrix, i, n)
        n -= 2
    return


def rotate_matrix_wrapper(square_matrix):
    rotate_matrix(square_matrix)
    return square_matrix


if __name__ == "__main__":
    exit(
        generic_test.generic_test_main(
            "matrix_rotation.py", "matrix_rotation.tsv", rotate_matrix_wrapper
        )
    )

File: epi_judge_cpp/matrix_rotation.cc
#include <tuple>
#include <vector>

#include "test_framework/generic_test.h"
using std::vector;
void RotateMatrix(vector<vector<int>> *square_matrix_ptr) {
  auto &sm{*square_matrix_ptr};
  if (sm.size() < 2) {
    return;
  }
  auto top = std::make_tuple(0, sm.size() - 2, 0);
  auto right = std::make_tuple(0, sm.size() - 2, sm.size() - 1);
  auto bottom = std::make_tuple(sm.size() - 1, 1, sm.size() - 1);
  auto left = std::make_tuple(sm.size() - 1, 1, 0);

  vector<int> swap(sm.size(), 0);
  for (auto count = 0; count < sm.size() / 2; count++) {
    auto [cbegin, cend, row] = top;
    for (auto col = cbegin, i = 0; col <= cend; col++, i++) {
      swap[i] = sm[row][col];
    }
    {
      auto [rbegin, rend, col] = right;
      for (auto row = rbegin, i = 0; row <= rend; row++, i++) {
        std::swap(swap[i], sm[row][col]);
      }
      right = {rbegin + 1, rend - 1, col - 1};
    }
    {
      auto [cbegin, cend, row] = bottom;
      for (int col = cbegin, i = 0; col >= cend; col--, i++) {
        std::swap(swap[i], sm[row][col]);
      }
      bottom = {cbegin - 1, cend + 1, row - 1};
    }
    {
      auto [rbegin, rend, col] = left;
      for (int row = rbegin, i = 0; row >= rend; row--, i++) {
        std::swap(swap[i], sm[row][col]);
      }
      left = {rbegin - 1, rend + 1, col + 1};
    }
    {
      auto [cbegin, cend, row] = top;
      for (auto col = cbegin, i = 0; col <= cend; col++, i++) {
        std::swap(swap[i], sm[row][col]);
      }
      top = {cbegin + 1, cend - 1, row + 1};
    }
  }

  return;
}
vector<vector<int>> RotateMatrixWrapper(vector<vector<int>> square_matrix) {
  RotateMatrix(&square_matrix);
  return square_matrix;
}

int main(int argc, char *argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"square_matrix"};
  return GenericTestMain(args, "matrix_rotation.cc", "matrix_rotation.tsv",
                         &RotateMatrixWrapper, DefaultComparator{},
                         param_names);
}
